credential/cargo-credential/src/error.rs

```diff
@@ -0,0 +1,193 @@
+use serde::{Deserialize, Serialize};
+use std::error::Error as StdError;
+use thiserror::Error as ThisError;
+
+/// Credential provider error type.
+/// 
+/// `UrlNotSupported` and `NotFound` errors both cause Cargo
+/// to attempt another provider, if one is available. The other 
+/// variants are fatal.
+/// 
+/// Note: Do not add a tuple variant, as it cannot be serialized.
+#[derive(Serialize, Deserialize, ThisError, Debug)]
+#[serde(rename_all = "kebab-case", tag = "kind")]
+#[non_exhaustive]
+pub enum Error {
```

Comment:
For Deserialization, do we need to worry about older credential providers having more `Error` variants?

Reply:
I think we're OK making a breaking change by removing variants here since it's unstable and it was only added last week.

Reply:
I meant for when this becomes stable, what constraints are we putting on evolution and how do we reduce them.

Reply:
Great point! I added a new `Unknown` variant marked with `serde(other)` to each of the enums that's serialized across the JSON protocol. That enables deserialization to succeed even if there are new variants added. Between this and marking the enums `non_exhaustive`, we're hopefully flexible enough.